flamingo/core/templating/jinja2.py

```diff
@@ -38,6 +39,14 @@
 
 logger = logging.getLogger('flamingo.core.templating.Jinja2')
 
+# starting from jinja2 3.0.0, we have to use pass_context instead of
+# contextfunction
+
+if hasattr(jinja2, "pass_context"):
+    pass_context = jinja2.pass_context
+else:
+    pass_context = jinja2.contextfunction
```

Comment:
To fix the codecov warning, you might do:

    try:
        from jinja2 import contextfunction
    except ImportError
        from jinja2 import pass_context as contextfunction

Reply:
Another alternative could be to drop jinja2 < 3 support.

Reply:
In 3.0.x, contextfunction has been deprecated and pass_context does already exist. Starting from 3.1.0, contextfunction has been removed. I'll improve the description.

Reply:
The try...except variant above has the disadvantage that it uses the deprecated contextversion in jinja2 3.0.x. I think I leave it as it is.

Reply:
Well, up to now the code uses `contextfunction`, so that seem natural to me. If you want to convert to `pass_context`, just do the construct the other way round. IMHO a try on an import is more pythonic than importing the whole module and doing hasattr on it.

Reply:
I tried this, but while testing I found out that we get other errors (from markup_safe) when testing with jinja2 < 3.0.0. So I'm abandoning this variant now completely and require >= 3.0.0.

Reply:
It seems we don't have any need to stay compatible with jinja2 < 3, so I prefer this simpler solution.